### PR TITLE
BUG: Fix pickling of scalars with NPY_LISTPICKLE 

### PR DESCRIPTION
**Pull request by seberg**
Structured void scalars (or potentially other dtypes using LISTPICKLE)
were incorrectly pickled and unpickled previously.
While the pickled value was valid and could have been used to
reconstruct the data, the reconstruction itself would always fail.

The approach here is to always pickle the full array and use the
arrays pickle/unpickle machinery which handles structured void
correctly.  Future work could improve on this probably.

In the (hopefully impossible) case that someone tries to unpickle
a numpy object scalar, the code is now fixed to return the pickled
value unmodified (but INCREF'd), however a deprecation-warning
is given, since such an object should never be created.

--- 

The second commit adds a deprecation warning (just in case, it should not be possible to reach it).  The first commit should also be backportable.

This was previously looked into in gh-14940, which fixed the worst of it, but did not actually end up in useful behaviour.  Here its fixed by pickling the array, which may not be ideal, but it is better than buggy behaviour.

### CODE REVIEW

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 1]
```diff
@@ -2003,20 +2003,40 @@ array_scalar(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwds)
...
     }
     if (PyDataType_FLAGCHK(typecode, NPY_LIST_PICKLE)) {
-        if (!PySequence_Check(obj)) {
-            PyErr_SetString(PyExc_TypeError,
-                            "found non-sequence while unpickling scalar with "
-                            "NPY_LIST_PICKLE set");
+        if (typecode->type_num == NPY_OBJECT) {
+            if (DEPRECATE(
+                    "Attempting to unpickling a numpy object scalar; "
+                    "object scalars should never be created. If this was a "
```
Comment (mattip) [inline numpy/core/src/multiarray/multiarraymodule.c:2016]:
```suggestion
            /* Deprecated 2020-11-24, NumPy 1.20 */
            if (DEPRECATE(
                    "Unpickling a scalar with object dtype is deprecated. "
                    "Object scalars should never be created. If this was a "
```

**Review by mattip — APPROVED** [review]

File: numpy/core/src/multiarray/multiarraymodule.c (hunk 1 above)
Comment (seberg) [reply numpy/core/src/multiarray/multiarraymodule.c]: Thanks, updated. I could push out the commit to make backporting easier if we do not want to include it in a possible backport.

**Comment by mattip** [comment]
marking for 1.20, feel free to remove the milestone if you disagree

**Comment by charris** [comment]
If this isn't a regression I'll skip the backport.

**Comment by charris** [comment]
Thanks Sebastian.

**Pull request merged**